### DISCUSSION TITLE
Install a package on specific platforms only

### DISCUSSION THREAD

**Discussion by avivanoff** [microsoft/vcpkg#37150] [General]
Let's say I have the following vcpkg manifest:
```
{
  "dependencies": [
    {
      "name": "nlohmann-json",
      "version>=": "3.11.3"
    }
  ]
}
```
Is there a way to specify that I only want this dependency on Cent 7?

**Comment by autoantwort** [comment]
There are `platform` expressions, but CentOS is not a known platform: https://learn.microsoft.com/en-us/vcpkg/reference/vcpkg-json#platform-expression

**Reply by avivanoff** [reply]
I know about platform expressions. But as you pointed out there is no way to specify a specific distribution and/or version. Any other options?

**Comment by dg0yt** [comment]
I assume that 
- your software depends on nlohmann-json for more platforms than just "Cent 7",
- you actually want to use a system-provided nlohmann-json on platforms other than "Cent 7".

In this case you need to actively provide an empty (or minimal) overlay port for the general nlohmann-json dependency.

**Reply by avivanoff** [reply]
Could you elaborate on "empty (or minimal) overlay port"? My knowledge of vcpkg is not that advanced.

**Reply by dg0yt** [reply]
"Overlay ports" are separate from the main registry and added to the build via `--overlay-ports=<dir>`. Cf.  
https://learn.microsoft.com/en-us/vcpkg/consume/install-locally-modified-package
The key is to provide the overlay ports options on the platforms where you need it (i.e. *not*  on "Cent 7").

"Empty" means that the port doesn't install anything but just satisfies the vcpkg dependency resolution. It needs a `vcpkg.json` to declare the name, version etc., and a `portfile.cmake` with just `set(VCPKG_POLICY_EMPTY_PACKAGE enabled)`.
In some cases, ports also carry wrappers or similar modifications of downstream behavior. Then an empty port might not be enough - you might need to add a "minimal" subset of the adapters.

**Reply by avivanoff** [reply]
Since we are building from the same source on multiple platforms, can the overlay port be configured to provide empty/minimal based on the target platform name/version?

**Reply by dg0yt** [reply]
It is up to your scripting, including CMake scripting before the `project` command - that's the point where the manifest mode installation takes place.
